fms/utils/generation.py

```diff
@@ -20,9 +20,73 @@ def _make_cache_contiguous(past_key_value_states):
     return n_kv_s
 
 
+def left_pad(input_ids: torch.Tensor, max_length: int, pad_id: int = 0) -> torch.Tensor:
+    """
+    left pad an input_ids tensor
+
+    Parameters
+    ----------
+    input_ids: torch.Tensor
+        input ids corresponding to a single sequence in a batch
+    max_length: int
+        the max length to pad to
+    pad_id: int
+        the token to set as a pad in the resulting tensor
+
+    Returns
+    -------
+    torch.Tensor
+        a left padded tensor
+    """
+    pads_tensor = torch.tensor(
+        [pad_id] * (max_length - input_ids.size(0)),
+        device=input_ids.device,
+        dtype=torch.long,
+    )
+    return torch.cat((pads_tensor, input_ids))
+
+
+def __create_attention_mask(
+    input_ids: torch.Tensor, use_cache: bool, is_prompt: bool, pad_id: int = -1
+) -> Optional[torch.Tensor]:
+    """
+    Optionally create an attention mask if one is required
+
+    Parameters
+    ----------
+    input_ids: torch.Tensor
```

Comment:
should we take a list of tensors and stack them here instead of in calling code?

Reply:
this gets called in each iteration of generate, we don't have the list of tensors anymore, just the batched torch.Tensor